### PR TITLE
remove idr-specific code

### PR DESCRIPTION
**Pull request by will-moore**
In order to use this for other (non-IDR) deployments, we need to remove anything that is specific to IDR:
To test:
 - Go to https://merge-ci.openmicroscopy.org/web/gallery/ (public user OR login first)
 - Should see new gallery UI without any IDR-specific features
 - On IDR ```test70```, with full IDR config and this PR deployed, IDR UI should behave as before with no changes.
 - Going to /gallery-settings/ will show the current gallery config JSON data for ```test70``` or at /gallery/gallery-settings/ for merge-ci (since gallery is not configured at root URL).

With this PR, we need only a single config to see the "new" gallery UI instead of the old one:
e.g.
```
$ bin/omero config set omero.web.gallery.category_queries '{"all":{"label":"All Studies", "index":0, "query":"FIRST50:Name"}}'
```
will list up to 50 Screens and Projects, sorted by Name in the first category.

Screen Shot 2019-06-26 at 14 34 02

To restore IDR functionality we now need the following configs (to be tested at https://github.com/IDR/deployment/pull/189)

```
$ bin/omero config set omero.web.gallery.favicon "https://idr.openmicroscopy.org/about/img/logos/favicon-idr.ico"

$ bin/omero config set omero.web.gallery.top_left_logo '{"src": "http://idr.openmicroscopy.org/about/img/logos/logo-idr.svg"}'

$ bin/omero config set omero.web.gallery.top_right_links '[{"text": "About", "href": "https://idr.openmicroscopy.org/about/index.html", "submenu":[{"text": "Overview", "href": "https://idr.openmicroscopy.org/about/index.html"}, {"text": "Published studies", "href": "https://idr.openmicroscopy.org/about/studies.html"}, {"text": "API Access", "href": "https://idr.openmicroscopy.org/about/api.html"}, {"text": "Data download", "href": "https://idr.openmicroscopy.org/about/download.html"}, {"text": "Image Tools Resource (ITR)", "href": "https://idr.openmicroscopy.org/about/itr.html"}, {"text": "Virtual Analysis Environment (VAE)", "href": "https://idr.openmicroscopy.org/jupyter"}, {"text": "Deployment", "href": "https://idr.openmicroscopy.org/about/deployment.html"}]},{"text": "Submissions", "href": "https://idr.openmicroscopy.org/about/submission.html", "submenu": [{"text": "Overview", "href": "https://idr.openmicroscopy.org/about/submission.html"}, {"text": "Screens", "href": "https://idr.openmicroscopy.org/about/screens.html"},        {"text": "Experiments", "href": "https://idr.openmicroscopy.org/about/experiments.html"}]}]'

$ bin/omero config set  omero.web.gallery.subheading_html "The Image Data Resource (IDR) is a public repository of image datasets from published scientific studies,<br/>where the community can submit, search and access high-quality bio-image data."

$ bin/omero config set omero.web.gallery.footer_html IDR

// Existing settings...
$ bin/omero config set omero.web.gallery.filter_keys '[{"label": "Name (IDR number)", "value": "Name"}, "Imaging Method", "License", "Organism", "Publication Authors", "Publication Title", "Screen Technology Type", "Screen Type", "Study Type"]'

$ bin/ omero config set omero.web.gallery.base_url "https://idr.openmicroscopy.org/"

$ bin/omero config set omero.web.gallery.category_queries '{"latest": {"label": "Most Recent studies", "index": 0, "query": "LAST10:date"}, "timelapse": {"label": "Time-lapse imaging", "index": 1, "query": "Study Type:time OR Study Type:5D OR Study Type:3D-tracking"}, "lightsheet": {"label": "Light sheet fluorescence microscopy", "index": 2, "query": "Study Type:light sheet"}, "proteinlocalization": {"label": "Protein localization studies", "index": 3, "query": "Study Type:protein localization"}, "histology": {"label": "Digital pathology imaging", "index": 4, "query":"Study Type:histology"}, "yeast": {"label": "Yeast studies", "index": 5, "query": "Organism: Saccharomyces cerevisiae OR Organism:Schizosaccharomyces pombe"}, "humancellscreen": {"label": "High-content screening (human)", "index": 6, "query": "Organism:Homo sapiens AND Study Type:high content screen"}}'

$ bin/omero config set omero.web.gallery.study_short_name '[{"key": "Name", "regex": "^(.*?)-.*?(.)$", "template": "$1$2"}]'
```

### CODE REVIEW

**Comment by joshmoore** [comment]
> let idrId = studyData.Name.split('-')[0] + studyData.Name[studyData.Name.length-1];  // idr0001A

@will-moore : how about a regex? In this case, something like `^(.*?)-.*?(.)$`

**Comment by mtbc** [comment]
On https://merge-ci.openmicroscopy.org/web/gallery/ seeing weird terms like "confocal imagess" and "HeLa cells in mitosiss", wonder why the extra 's'.

**Comment by mtbc** [comment]
Would be lovely if it could wrap the studies, all six are in a horizontal scrolly thing for me even after I maximize the window.

**Comment by will-moore** [comment]
To test this PR with IDR settings (added in https://github.com/IDR/deployment/pull/189) please try the IDR test server. It should appear unchanged compared with idr.openmicroscopy.org
Check:
 - favicon
 - top-left icon and link to homepage
 - top-right toolbar menus - links all working?
 - Header title "Welcome to IDR"
 - subheader text below the header
 - page footer

**Comment by joshmoore** [comment]
@will-moore : I'm missing a search bar on osx/chrome.

**Comment by mtbc** [comment]
Regarding https://github.com/ome/omero-gallery/pull/42#issuecomment-506351011 all looking good to me. :+1:

**Comment by will-moore** [comment]
Will - discussed at IDR meeting:
 - Try regex for ID/name. Use 'Project: 1' by default.
 - If only 1 category configured - wrap studies onto multiple rows (no horizontal scrollbar).

Also:
 - Document other 'hidden' IDR-specific behaviour (e.g. parsing of Description)
 - Write docs for all existing configs and update README screenshots etc.

**Comment by joshmoore** [comment]
```
./omero_gallery/gallery_settings.py:128:1: W293 blank line contains whitespace
```

**Comment by will-moore** [comment]
Ready for testing again at https://merge-ci.openmicroscopy.org/web/gallery/
 - The 'short name' above each study is now 'Project: 1' by default
 - If there is only 1 'category' then we wrap the study thumbnails into multiple rows (no scrollbar). This is configured at https://merge-ci.openmicroscopy.org/jenkins/job/OMERO-web/configure to sort by Name
```
bin/omero config set omero.web.gallery.category_queries '{"all":{"label":"All Studies", "index":0, "query":"FIRST50:Name"}}'
```
NB: Some of the Studies in the merge-ci example above have a Map Annotation with "Publication Title" which is used instead of Name in the UI, so they don't appear to be sorted by Name (but they are).

**Comment by joshmoore** [comment]
Minor CSS issue on small windows where the rows aren't centered:
Screen Shot 2019-07-02 at 11 51 36

File: omero_gallery/static/gallery/model.js [hunk 1]
```diff
@@ -509,6 +517,53 @@ StudiesModel.prototype.getStudyImage = function getStudyImage(obj_type, obj_id,
...
+      if (!value) {
+        continue;
+      }
+
+      if (STUDY_SHORT_NAME[i]['regex']) {
+        var re = new RegExp(STUDY_SHORT_NAME[i]['regex']);
+        var match = re.exec(value);
+
+        if (match && match.length > 1) {
+          newShortName = match.slice(1).join("");
```
Comment (will-moore) [inline omero_gallery/static/gallery/model.js]: @joshmoore So the regex ```^(.*?)-.*?(.)$``` always returns a list of 3 matches: the full name "idr0010-doil-dnadamage/screenA", "idr0010", "A". So to get the matches I want, I have to ignore the first item in the list.
I tried a bunch of different regexes to try and only match the sub-parts of the name but no joy. So I'm hoping that any other regex we want to use would behave in the same way, but I'm not convinced?
The config allows you to apply the regex to E.g. Description or any other Map-Annotation Value (or just to use a Map-Annotation value directly), so we may not need the regex if we add the 'short name' to the Key-Value pairs. cc @sbesson @manics

Comment (joshmoore) [reply omero_gallery/static/gallery/model.js]: You're right. The capturing groups aren't going to be sufficiently flexible. We'd need to _either_ say, "there shall *always* be exactly two sets of parens" or we'd need to use capturing group names like [import targets](https://docs.openmicroscopy.org/omero/5.4.10/users/cli/import-target.html#importing-using-regular-expressions). But against, not flexible enough. If that's the case, this should probably take both the regex as well as the substitution string. 

see https://ant.apache.org/manual/Tasks/replaceregexp.html

```
               match="^(.*?)-.*?(.)$"
               replace="\1\2"
``

which in Python land would be:

```
pat = re.compile("^(.*?)-.*?(.)$")
pat.sub(r"\g<1>\g<2>", string1)
```

Comment (will-moore) [reply omero_gallery/static/gallery/model.js]: It feels like this is getting over-complicated, when we don't really know of a use-case yet (other than IDR). Needs more design / UI discussion first.

**Comment by will-moore** [comment]
@joshmoore the off-centre 2-column layout is the same as at eg. http://idr.openmicroscopy.org/search/?query=Organism:Homo%20sapiens when resizing window.
Needs a bit more work to fix that.

File: omero_gallery/static/gallery/model.js (hunk 1 above)
Comment (will-moore) [reply omero_gallery/static/gallery/model.js]: OK, so it wasn't much more work to support templates like "{{1}}{{2}}".

File: omero_gallery/static/gallery/model.js [hunk 2]
```diff
@@ -509,6 +517,61 @@ StudiesModel.prototype.getStudyImage = function getStudyImage(obj_type, obj_id,
...
+        value = model.getStudyValue(study, key);
+      }
+
+      if (!value) {
+        continue;
+      }
+
+      if (STUDY_SHORT_NAME[i]['regex'] && STUDY_SHORT_NAME[i]['template']) {
+        var re = new RegExp(STUDY_SHORT_NAME[i]['regex']);
+        var groups = re.exec(value);
```
Comment (joshmoore) [inline omero_gallery/static/gallery/model.js]: I'd still like to propose one more time just using straight `replace()` syntax here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace

Comment (will-moore) [reply omero_gallery/static/gallery/model.js]: @joshmoore Sorry for not twigging sooner. JS ```replace()``` with regex works pretty nicely.
Done in a354e4a. Will now update the config PR https://github.com/IDR/deployment/pull/190

**Comment by will-moore** [comment]
Finally deployed with latest settings on test70 and all looks good!

File: omero_gallery/templates/webgallery/categories/base.html [hunk 3]
```diff
@@ -6,13 +6,17 @@
         <meta http-equiv="x-ua-compatible" content="ie=edge">
         <meta name="viewport" content="width=device-width, initial-scale=1.0">
         <title>IDR: Image Data Resource</title>
```
Comment (joshmoore) [inline omero_gallery/templates/webgallery/categories/base.html]: Oops. I missed this one.

File: omero_gallery/templates/webgallery/categories/base.html [hunk 4]
```diff
@@ -84,8 +92,10 @@
 
         <hr class="whitespace">
 
+        {% if footer_html %}
         <div class="callout large secondary">
             <div class="row">
+                {% if footer_html == 'IDR' %}
```
Comment (joshmoore) [inline omero_gallery/templates/webgallery/categories/base.html]: Slightly odd to have this specified here, but probably not an issue for now.

**Comment by manics** [comment]
Searching for an image attribute on test70 results in an `undefined` Study ID: https://test70/cell/search/?query=mapr_gene:INCENP
Screen Shot 2019-08-27 at 17 36 44

Also seen some weird behaviour when searching for phenotypes, see private screen recording

File: omero_gallery/templates/webgallery/categories/base.html (hunk 4 above)
Comment (will-moore) [reply omero_gallery/templates/webgallery/categories/base.html]: This allows you to specify ```footer_hml = "<div>My footer</div>"``` in the config, but for IDR it's not practicable to put all the footer html we need in the config. So in this once case, that IDR html will remain in the gallery code and simply be enabled for IDR using the ```footer_html == 'IDR'``` flag.

**Comment by will-moore** [comment]
To test fixes:
 - https://test70/cell/search/?query=mapr_gene:INCENP fixed "undefined" links
 - <title> should be same.
 - https://merge-ci.openmicroscopy.org/web/gallery/ <title> should now be "OMERO.gallery"

**Comment by pwalczysko** [comment]
We discussed with @will-moore about how would this fare when there is no public user on the server at all (such as outreach). In discussion with @joshmoore we concluded that an immediate redirect to login page should be the behaviour in such case. Is that here ? Or is it something for another PR ?

**Comment by will-moore** [comment]
I've added the login in a different PR: https://github.com/ome/omero-gallery/pull/45

**Comment by pwalczysko** [comment]
No problems here for me then.

**Comment by manics** [comment]
Released as https://github.com/ome/omero-gallery/releases/tag/v3.2.0a8

**Pull request merged**